src/digital/v2.rs

```diff
@@ -136,3 +136,27 @@ pub trait InputPin {
     /// Is the input pin low?
     fn is_low(&self) -> Result<bool, Self::Error>;
 }
+
+/// Describe pin that can switch its input/ouput mode on fly.
+/// Should be used to communicate with custom protocols based hardware.
+///
+/// *This trait is available if embedded-hal is built with the `"unproven"` feature.*
+#[cfg(feature = "unproven")]
+pub trait InputOutputPin<Error>: InputPin<Error = Error> + OutputPin<Error = Error> {
+    /// Switch pin into output mode,
+    /// 
+    /// Result::Ok value is true if pin mode was changed or false if pin already was in output mode.
+    ///
+    /// Expected behaviour after success result:
+    /// - OutputPin functions should work as for ouptut pin
+    /// - InputPin functions should return errors
```

Comment:
This cannot work because an object implementing `InputPin` cannot know that it also is implementing `InputOutputPin` and needs to behave differently if it is in output mode and vice versa. That's what I mean by "this is not composable".

Reply:
But there can be some kind of structure which does know about current pin mode and can use one or another nested pin structure (via enum for example) to provide required functionality.

Reply:
You cannot retroactively impose additional restrictions or interfaces on existing implementations with a new trait.

Reply:
Why not? HAL has not released yet and it has unproven feature. So it is possible to add new unproven trait to new HAL versions.

Reply:
Sure, but whatever you put in the traits will not have any influence on the trait impls which are out there.